### ISSUE TITLE
Cannot destroy '<pool>': dataset already exists

### ISSUE THREAD

**Issue by pakjebakmeel** [zfsonlinux/zfs#4196]
Hi all,

I noticed that my zfs-auto-snapshot was returning errors, it seems something is quite wrong. When trying to destroy a ZVOL:

>gentoo / # zfs destroy storage/vm/john -r
>cannot destroy 'storage/vm/john': dataset already exists

I have 2 pools, a mirror based on 2 slices and a RAIDZ2 pool consisting of 6 3TB disks:

>gentoo / # zpool status
>pool: storage
>state: ONLINE
>scan: scrub repaired 0 in 8h13m with 0 errors on Sat Jan  9 22:30:45 2016
>config:
>
>NAME                                          STATE     READ WRITE CKSUM
>storage                                       ONLINE       0     0     0
>raidz2-0                                    ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T0999727  ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1035352  ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1157517  ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1248064  ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1429505  ONLINE       0     0     0
>ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1625883  ONLINE       0     0     0
> 
>errors: No known data errors
>
>pool: work
>state: ONLINE
>scan: scrub repaired 0 in 6h11m with 0 errors on Fri Jan  1 10:11:08 2016
>config:
>
>NAME                                                STATE     READ WRITE CKSUM
>work                                                ONLINE       0     0     0
>mirror-0                                          ONLINE       0     0     0
>ata-WDC_WD20EFRX-68EUZN0_WD-WMC4M1460778-part4  ONLINE       0     0     0
>ata-WDC_WD20EFRX-68EUZN0_WD-WMC4M1808458-part4  ONLINE       0     0     0
>
>errors: No known data errors

ZFS List output:

>gentoo / # zfs list
>NAME                     USED  AVAIL  REFER  MOUNTPOINT
>storage                 6.31T  4.18T  5.20T  /mnt/storage
>storage/vm               281G  4.18T   192K  none
>storage/vm/esxi          168G  4.18T  29.0G  /mnt/esxi
>storage/vm/john          112G  4.18T   192K  none
>work                     576G  1.10T  21.2G  /mnt/work
>work/vm                  250G  1.10T    96K  none
>work/vm/ank1ap02         103G  1.10T    96K  none
>work/vm/ank1ap02/dsk00   103G  1.19T  10.0G  -
>work/vm/ank1dc01        43.4G  1.10T    96K  none
>work/vm/ank1dc01/dsk00  43.4G  1.13T  10.6G  -
>work/vm/ank1us02         103G  1.10T    96K  none
>work/vm/ank1us02/dsk00   103G  1.17T  38.3G  -

When trying to destroy any snapshot regardless on which pool results in:

gentoo / # zfs destroy storage/vm/john/dsk00@zfs-auto-snap_02-2015-10-31-0230
internal error: out of memory

Same for creating snapshots, also the internal error: out of memory. The userland tools I always merge together with a new kernel modules so that is up-to-date. I'm currently running:

>gentoo / # uname -a
>Linux gentoo 4.1.12-gentoo #6 SMP Sat Dec 5 14:45:52 CET 2015 x86_64 Intel(R) Xeon(R) CPU >E3-1230 v3 @ 3.30GHz GenuineIntel GNU/Linux

>gentoo / # modinfo zfs
>filename:       /lib/modules/4.1.12-gentoo/extra/zfs/zfs.ko
>version:        0.6.5.3-r2-gentoo
>license:        CDDL
>author:         OpenZFS on Linux
>description:    ZFS
>srcversion:     9258B85484EEFAA7F78E055
>depends:        spl,znvpair,zcommon,zunicode,zavl
>vermagic:       4.1.12-gentoo SMP mod_unload

>gentoo / # modinfo spl
>filename:       /lib/modules/4.1.12-gentoo/extra/spl/spl.ko
>version:        0.6.5.3-r1-gentoo
>license:        GPL
>author:         OpenZFS on Linux
>description:    Solaris Porting Layer
>srcversion:     EE741F18CAD1766B535EC0D
>depends:
>vermagic:       4.1.12-gentoo SMP mod_unload

ZDB is also showing some scary message about metaslabs:

>gentoo / # zdb storage -m
>
>Metaslabs:
>        vdev          0
>        metaslabs   130   offset                spacemap          free
>        ---------------   -------------------   ---------------   -------------
>        metaslab      0   offset            0   spacemap     38   free    61.2M
>        metaslab      1   offset   2000000000   spacemap     45   free    7.34M
>        metaslab      2   offset   4000000000   spacemap     85   free    34.2G
>        metaslab      3   offset   6000000000   spacemap    121   free    62.6M
>        metaslab      4   offset   8000000000   spacemap    163   free    54.8M
>        metaslab      5   offset   a000000000   spacemap    164   free    29.6M
>        metaslab      6   offset   c000000000   spacemap    188   free    81.9M
>        metaslab      7   offset   e000000000   spacemap    210   free     121M
>        metaslab      8   offset  10000000000   spacemap    232   free     113M
>        metaslab      9   offset  12000000000   spacemap    234   free     171M
>        metaslab     10   offset  14000000000   spacemap    302   free    65.8M
>        metaslab     11   offset  16000000000   spacemap    307   free    28.6M
>        metaslab     12   offset  18000000000   spacemap    330   free    41.6M
>        metaslab     13   offset  1a000000000   spacemap    331   free    96.8M
>        metaslab     14   offset  1c000000000   spacemap    354   free    53.9M
>        metaslab     15   offset  1e000000000   spacemap    355   free    79.3M
>        metaslab     16   offset  20000000000   spacemap    398   free    63.6M
>        metaslab     17   offset  22000000000   spacemap    449   free    35.7M
>        metaslab     18   offset  24000000000   spacemap    452   free    38.1M
>        metaslab     19   offset  26000000000   spacemap    475   free    86.6M
>        metaslab     20   offset  28000000000   spacemap    476   free    78.8M
>        metaslab     21   offset  2a000000000   spacemap    499   free    60.7M
>        metaslab     22   offset  2c000000000   spacemap    501   free    90.7M
>        metaslab     23   offset  2e000000000   spacemap    545   free    59.8M
>        metaslab     24   offset  30000000000   spacemap    546   free    43.9M
>        metaslab     25   offset  32000000000   spacemap     37   free     721M
>        metaslab     26   offset  34000000000   spacemap    573   free    38.9M
>        metaslab     27   offset  36000000000   spacemap    574   free    21.7M
>        metaslab     28   offset  38000000000   spacemap    598   free    61.0M
>        metaslab     29   offset  3a000000000   spacemap    621   free    85.6M
>        metaslab     30   offset  3c000000000   spacemap    622   free    81.4M
>        metaslab     31   offset  3e000000000   spacemap    667   free    60.3M
>        metaslab     32   offset  40000000000   spacemap    668   free    56.0M
>        metaslab     33   offset  42000000000   spacemap    694   free    1.64G
>        metaslab     34   offset  44000000000   spacemap    857   free    69.0G
>        metaslab     35   offset  46000000000   spacemap    301   free     508M
>        metaslab     36   offset  48000000000   spacemap   1244   free    11.2G
>        metaslab     37   offset  4a000000000   spacemap   1359   free     167M
>        metaslab     38   offset  4c000000000   spacemap   1552   free     115M
>        metaslab     39   offset  4e000000000   spacemap   1731   free    1.00G
>        metaslab     40   offset  50000000000   spacemap   1905   free    92.4M
>        metaslab     41   offset  52000000000   spacemap   2017   free     103M
>        metaslab     42   offset  54000000000   spacemap    448   free    10.5G
>        metaslab     43   offset  56000000000   spacemap   2191   free     136M
>        metaslab     44   offset  58000000000   spacemap   2337   free    3.13G
>        metaslab     45   offset  5a000000000   spacemap   2462   free    35.5G
>        metaslab     46   offset  5c000000000   spacemap   2489   free    17.5G
>        metaslab     47   offset  5e000000000   spacemap   2564   free    11.5G
>        metaslab     48   offset  60000000000   spacemap   2595   free    16.6G
>        metaslab     49   offset  62000000000   spacemap     23   free    1.66G
>        metaslab     50   offset  64000000000   spacemap     36   free    88.7M
>        metaslab     51   offset  66000000000   spacemap    572   free    1.21G
>        metaslab     52   offset  68000000000   spacemap   4951   free    5.66G
>        metaslab     53   offset  6a000000000   spacemap   5528   free     537M
>        metaslab     54   offset  6c000000000   spacemap   6134   free    30.2G
>        metaslab     55   offset  6e000000000   spacemap  15916   free    98.4G
>        metaslab     56   offset  70000000000   spacemap     39   free    25.8G
>        metaslab     57   offset  72000000000   spacemap   4097   free     117G
>        metaslab     58   offset  74000000000   spacemap  24584   free     109M
>        metaslab     59   offset  76000000000   spacemap  28673   free    51.0G
>        metaslab     60   offset  78000000000   spacemap    300   free    2.59G
>        metaslab     61   offset  7a000000000   spacemap  12289   free     105G
>        metaslab     62   offset  7c000000000   spacemap   4100   free     323M
>        metaslab     63   offset  7e000000000   spacemap  12292   free     133M
>        metaslab     64   offset  80000000000   spacemap     40   free     413M
>        metaslab     65   offset  82000000000   spacemap   1459   free    74.7G
>        metaslab     66   offset  84000000000   spacemap     53   free     113G
>        metaslab     67   offset  86000000000   spacemap     68   free     119M
>        metaslab     68   offset  88000000000   spacemap   4099   free     512M
>        metaslab     69   offset  8a000000000   spacemap   2336   free     592M
>        metaslab     70   offset  8c000000000   spacemap   2461   free    2.98G
>        metaslab     71   offset  8e000000000   spacemap    969   free    25.8M
>        metaslab     72   offset  90000000000   spacemap   1163   free    2.32G
>        metaslab     73   offset  92000000000   spacemap   1170   free     130M
>        metaslab     74   offset  94000000000   spacemap   1173   free    34.7G
>        metaslab     75   offset  96000000000   spacemap    233   free    1.02G
>        metaslab     76   offset  98000000000   spacemap   4113   free     110G
>        metaslab     77   offset  9a000000000   spacemap     51   free     103G
>        metaslab     78   offset  9c000000000   spacemap   4107   free     437M
>        metaslab     79   offset  9e000000000   spacemap   1119   free     226M
>        metaslab     80   offset  a0000000000   spacemap   1161   free    61.2G
>        metaslab     81   offset  a2000000000   spacemap   4108   free    1.27G
>        metaslab     82   offset  a4000000000   spacemap   4170   free     105M
>        metaslab     83   offset  a6000000000   spacemap  24583   free     126G
>        metaslab     84   offset  a8000000000   spacemap    963   free    51.6G
>        metaslab     85   offset  aa000000000   spacemap    418   free     127G
>        metaslab     86   offset  ac000000000   spacemap     52   free     127G
>        metaslab     87   offset  ae000000000   spacemap      0   free     128G
>        metaslab     88   offset  b0000000000   spacemap      0   free     128G
>        metaslab     89   offset  b2000000000   spacemap      0   free     128G
>        metaslab     90   offset  b4000000000   spacemap      0   free     128G
>        metaslab     91   offset  b6000000000   spacemap      0   free     128G
>        metaslab     92   offset  b8000000000   spacemap      0   free     128G
>        metaslab     93   offset  ba000000000   spacemap      0   free     128G
>        metaslab     94   offset  bc000000000   spacemap   2335   free     128G
>        metaslab     95   offset  be000000000   spacemap  24578   free     128G
>        metaslab     96   offset  c0000000000   spacemap      0   free     128G
>        metaslab     97   offset  c2000000000   spacemap      0   free     128G
>        metaslab     98   offset  c4000000000   spacemap      0   free     128G
>        metaslab     99   offset  c6000000000   spacemap      0   free     128G
>        metaslab    100   offset  c8000000000   spacemap      0   free     128G
>        metaslab    101   offset  ca000000000   spacemap      0   free     128G
>        metaslab    102   offset  cc000000000   spacemap      0   free     128G
>        metaslab    103   offset  ce000000000   spacemap      0   free     128G
>        metaslab    104   offset  d0000000000   spacemap   1118   free     128G
>        metaslab    105   offset  d2000000000   spacemap      0   free     128G
>        metaslab    106   offset  d4000000000   spacemap      0   free     128G
>        metaslab    107   offset  d6000000000   spacemap      0   free     128G
>        metaslab    108   offset  d8000000000   spacemap  24582   free     128G
>        metaslab    109   offset  da000000000   spacemap   4098   free     128G
>        metaslab    110   offset  dc000000000   spacemap      0   free     128G
>        metaslab    111   offset  de000000000   spacemap      0   free     128G
>        metaslab    112   offset  e0000000000   spacemap      0   free     128G
>        metaslab    113   offset  e2000000000   spacemap      0   free     128G
>        metaslab    114   offset  e4000000000   spacemap      0   free     128G
>        metaslab    115   offset  e6000000000   spacemap      0   free     128G
>        metaslab    116   offset  e8000000000   spacemap      0   free     128G
>        metaslab    117   offset  ea000000000   spacemap      0   free     128G
>        metaslab    118   offset  ec000000000   spacemap      0   free     128G
>        metaslab    119   offset  ee000000000   spacemap      0   free     128G
>        metaslab    120   offset  f0000000000   spacemap      0   free     128G
>        metaslab    121   offset  f2000000000   spacemap      0   free     128G
>        metaslab    122   offset  f4000000000   spacemap      0   free     128G
>        metaslab    123   offset  f6000000000   spacemap      0   free     128G
>        metaslab    124   offset  f8000000000   spacemap      0   free     128G
>        metaslab    125   offset  fa000000000   spacemap      0   free     128G
>        metaslab    126   offset  fc000000000   spacemap      0   free     128G
>        metaslab    127   offset  fe000000000   spacemap      0   free     128G
>        metaslab    128   offset 100000000000   spacemap      0   free     128G
>        metaslab    129   offset 102000000000   spacemap      0   free     128G
>
>space map refcount mismatch: expected 152 != actual 92

Something I noticed is that a scrub says it scrubs 9+ TB of data whilst I have nowhere near that much data on the storage volume. I am currently using 5,3TB..

Any clue what could be going on here? I have not seen any other reports of this apart from people that were sending/receiving incremental snapshots back in 2009 and 2012. Those issues seem to have been fixed. I do not send/receive snapshots so I don't this those are applicable to my situation.

Any thoughts? Could it be gentoo specific?

So to sum up:

Create snapshot : Out of memory error.
Destroy snapshot: Out of memory error.
Create snapshot: Out of memory error.
Destroy ZVOL: dataset already exists
Scrub shows to much data.
Space map refcount mismatch.

**Comment by kernelOfTruth** [comment]
> space map refcount mismatch: expected 152 != actual 92

referencing

https://github.com/zfsonlinux/zfs/issues/3094 Spacemap Problem with Data Error

https://github.com/zfsonlinux/zfs/issues/3577#issuecomment-143968398 Data corruption in cluster environment that uses zfs and shared storage

How old is that pool, when was it created ?

**Comment by pakjebakmeel** [comment]
I'm having the same issues on both pools.

>2015-04-19.00:09:27 zpool create -f -o ashift=12 -m /mnt/storage storage raidz2 /dev/disk/by-id/ata->WDC_WD30EZRX-00DC0B0_WD-WMC1T0999727 /dev/disk/by-id/ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1035352 /dev/disk/by-id/ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1157517 /dev/disk/by-id/ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1248064 /dev/disk/by->id/ata-WDC_WD30EZRX-00DC0B0_WD-WMC1T1429505 /dev/disk/by-id/ata-WDC_WD30EZRX->00DC0B0_WD-WMC1T1625883

>2015-04-19.21:28:24 zpool create -m /mnt/work -o ashift=12 work mirror /dev/disk/by-id/ata-WDC_WD20EFRX-68EUZN0_WD-WMC4M1460778-part4 /dev/disk/by-id/ata-WDC_WD20EFRX-68EUZN0_WD-WMC4M1808458-part4

I noticed that all was fine up to a couple of weeks ago. The last thing I actually did was:

>2015-12-04.03:10:02 zfs snapshot -o com.sun:auto-snapshot-desc=- -r storage@zfs-auto-snap_01-2015-12-04-0210
>2015-12-04.03:10:07 zfs destroy -r storage@zfs-auto-snap_01-2015-11-03-0210
>2015-12-04.22:23:11 zpool scrub storage
>2015-12-05.03:10:18 zfs snapshot -o com.sun:auto-snapshot-desc=- -r storage@zfs-auto-snap_01-2015-12-05-0210
>2015-12-05.03:10:25 zfs destroy -r storage@zfs-auto-snap_01-2015-11-04-0210
>2015-12-05.04:20:05 zfs snapshot -o com.sun:auto-snapshot-desc=- -r storage@zfs-auto-snap_02-2015-12-05-0320
>2015-12-05.04:20:07 zfs destroy -r storage@zfs-auto-snap_02-2015-10-17-0220
>2015-12-05.12:39:47 zfs create -V 500G storage/vm/esxi
>2015-12-05.12:42:30 zfs destroy storage/vm/esxi
>2015-12-05.12:42:46 zfs create storage/vm/esxi
>2015-12-05.12:43:16 zfs set mountpoint=/mnt/esxi storage/vm/esxi
>2015-12-05.12:57:17 zfs set sharenfs=rw=@192.168.13.0/28 storage/vm/esxi
>2016-01-01.04:00:09 zpool scrub storage
>2016-01-09.14:17:05 zpool scrub storage

So you can see it was doing snapshots just fine until I created the NFS share. It seems that after that no more snapshots were made or deleted. Not sure if it's related..

**Comment by pakjebakmeel** [comment]
As soon as I stop my NFS exports if behaves differently:

>gentoo / # zfs destroy storage/vm/john
>cannot destroy 'storage/vm/john': filesystem has children
>use '-r' to destroy the following datasets:
>storage/vm/john@zfs-auto-snap_01-2016-01-11-0210

>gentoo / # zfs destroy storage/vm/john -r
>cannot destroy 'storage/vm/john': dataset already exists

>gentoo / # zfs destroy storage/vm/john@zfs-auto-snap_01-2016-01-11-0210
>could not find any snapshots to destroy; check snapshot names.

>gentoo / # zfs list -t snapshot | grep john@zfs-auto-snap_01-2016-01-11-0210
>gentoo / #

**Comment by pakjebakmeel** [comment]
When using 

sys-fs/zfs-0.6.5.3 

instead of 

sys-fs/zfs-0.6.5.3-r2

It is working again.. I can delete the snapshots again. I presume this means the regression is for the Gentoo maintainer as the issue is not in ZFS particularly?

**Comment by kernelOfTruth** [comment]
@pakjebakmeel yes,

referencing: https://github.com/zfsonlinux/zfs/issues/4204 sys-fs/zfs-0.6.5.3-r2 cannot destroy snapshot: internal error: out of memory

CC: @ryao

**Comment by behlendorf** [comment]
Duplicate of #4112
